In [13]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [14]:
matchup_data = pd.read_csv("matchup_features.csv")

matchup_data.head()

,gameId,season_year,home_win,home_team,home_last_5_win_pct,home_offensive_average,home_defensive_average,home_last_5_point_diff,away_season_year,away_home_win,away_team,away_last_5_win_pct,away_offensive_average,away_defensive_average,away_last_5_point_diff,win_pct_diff,offense_diff,defense_diff,point_diff_diff
0,22000083,2021,0,Atlanta Hawks,NaN,NaN,NaN,NaN,2021,0,Cleveland Cavaliers,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,22000095,2021,0,Atlanta Hawks,NaN,NaN,NaN,NaN,2021,0,New York Knicks,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,22000110,2021,0,Atlanta Hawks,NaN,NaN,NaN,NaN,2021,0,Charlotte Hornets,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,22000153,2021,1,Atlanta Hawks,0.2,102.4,104.0,-1.6,2021,1,Philadelphia 76ers,0.6,119.6,117.2,2.4,-0.4,-17.2,-13.2,-4.0
4,22000205,2021,1,Atlanta Hawks,0.2,101.8,107.4,-5.6,2021,1,Minnesota Timberwolves,0.2,111.6,117.8,-6.2,0.0,-9.8,-10.4,0.6


In [23]:
feature_cols = [
    "win_pct_diff",
    "offense_diff",
    "defense_diff",
    "point_diff_diff"
]

model_df = matchup_data[
    feature_cols + ["home_win", "season_year"]
].dropna()
model_df.head()

,win_pct_diff,offense_diff,defense_diff,point_diff_diff,home_win,season_year
3,-0.4,-17.2,-13.2,-4.0,1,2021
4,0.0,-9.8,-10.4,0.6,1,2021
5,0.0,-0.2,1.6,-1.8,1,2021
6,-0.4,-8.4,10.6,-19.0,1,2021
7,0.2,-5.8,-13.2,7.4,0,2021


creating X and Y

In [25]:
X = matchup_data[feature_cols]

y = matchup_data["home_win"] # our target variable

In [26]:
train_df = model_df[model_df["season_year"] <= 2024].copy()
test_df = model_df[model_df["season_year"] == 2025].copy()

X_train = train_df[feature_cols]
y_train = train_df["home_win"]

X_test = test_df[feature_cols]
y_test = test_df["home_win"]

print("Training shape:", X_train.shape)
print("Testing shape:", X_test.shape)

Training shape: (5736, 4)
Testing shape: (1387, 4)


5736 rows and 4 columns. in training data from 2021-2024. 1387 rows and 4 columns in test data from 2024-2025.

Now we will be trying different models.